File: backend/src/handlers/aliases.rs
use axum::{
    extract::{Path, Query, State},
    Json,
};
use serde::Deserialize;
use uuid::Uuid;

use crate::{
    db::{
        aliases::AliasRow,
        alias_targets::{AliasTargetRow, AliasTargetDetail},
        types::ApiType,
    },
    error::{AppError, AppResult},
    services::aliases,
    state::AppState,
};

#[derive(Debug, Deserialize)]
pub struct ListAliasesQuery {
    #[serde(default = "default_page")]
    pub page: i64,
    #[serde(default = "default_page_size")]
    pub page_size: i64,
}

fn default_page() -> i64 {
    1
}

fn default_page_size() -> i64 {
    20
}

pub async fn list_aliases(
    State(state): State<AppState>,
    Query(query): Query<ListAliasesQuery>,
) -> AppResult<Json<Vec<AliasRow>>> {
    let aliases = aliases::list_aliases(&state.pool, query.page, query.page_size).await?;
    Ok(Json(aliases))
}

#[derive(Debug, Deserialize)]
pub struct CreateAliasRequest {
    pub name: String,
}

pub async fn create_alias(
    State(state): State<AppState>,
    Json(payload): Json<CreateAliasRequest>,
) -> AppResult<Json<AliasRow>> {
    let alias = aliases::create_alias(
        &state.pool,
        payload.name,
    )
    .await?;
    Ok(Json(alias))
}

pub async fn get_alias(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> AppResult<Json<AliasRow>> {
    let alias = aliases::get_alias(&state.pool, id)
        .await?
        .ok_or(AppError::NotFound)?;
    Ok(Json(alias))
}

#[derive(Debug, Deserialize)]
pub struct UpdateAliasRequest {
    pub name: Option<String>,
    pub enabled: Option<bool>,
}

pub async fn update_alias(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(payload): Json<UpdateAliasRequest>,
) -> AppResult<Json<AliasRow>> {
    let alias = aliases::update_alias(
        &state.pool,
        id,
        payload.name,
        payload.enabled,
    )
    .await?
    .ok_or(AppError::NotFound)?;
    Ok(Json(alias))
}

pub async fn delete_alias(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> AppResult<()> {
    if aliases::delete_alias(&state.pool, id).await? {
        Ok(())
    } else {
        Err(AppError::NotFound)
    }
}

// Alias Targets

pub async fn list_alias_targets(
    State(state): State<AppState>,
    Path(alias_id): Path<Uuid>,
) -> AppResult<Json<Vec<AliasTargetRow>>> {
    let targets = aliases::fetch_alias_targets(&state.pool, alias_id).await?;
    Ok(Json(targets))
}

pub async fn list_alias_target_details(
    State(state): State<AppState>,
    Path(alias_id): Path<Uuid>,
) -> AppResult<Json<Vec<AliasTargetDetail>>> {
    let details = aliases::fetch_all_alias_target_details(&state.pool, alias_id).await?;
    Ok(Json(details))
}

#[derive(Debug, Deserialize)]
pub struct CreateAliasTargetRequest {
    pub provider_id: Uuid,
    pub model_id: Uuid,
}

pub async fn create_alias_target(
    State(state): State<AppState>,
    Path(alias_id): Path<Uuid>,
    Json(payload): Json<CreateAliasTargetRequest>,
) -> AppResult<Json<AliasTargetRow>> {
    let target = aliases::create_alias_target(
        &state.pool,
        alias_id,
        payload.provider_id,
        payload.model_id,
    )
    .await?;
    Ok(Json(target))
}

#[derive(Debug, Deserialize)]
pub struct UpdateAliasTargetRequest {
    pub enabled: Option<bool>,
}

pub async fn update_alias_target(
    State(state): State<AppState>,
    Path((_alias_id, target_id)): Path<(Uuid, Uuid)>,
    Json(payload): Json<UpdateAliasTargetRequest>,
) -> AppResult<Json<AliasTargetRow>> {
    let target = aliases::update_alias_target(
        &state.pool,
        target_id,
        payload.enabled,
    )
    .await?
    .ok_or(AppError::NotFound)?;
    Ok(Json(target))
}

pub async fn delete_alias_target(
    State(state): State<AppState>,
    Path((_alias_id, target_id)): Path<(Uuid, Uuid)>,
) -> AppResult<()> {
    if aliases::delete_alias_target(&state.pool, target_id).await? {
        Ok(())
    } else {
        Err(AppError::NotFound)
    }
}

File: frontend/src/api/aliases.ts
import { requestJson } from "./client";

export interface Alias {
  id: string;
  name: string;
  enabled: boolean;
  created_at: string;
}

export interface AliasTarget {
  id: string;
  alias_id: string;
  provider_id: string;
  model_id: string;
  enabled: boolean;
  created_at: string;
}

export interface AliasTargetDetail extends AliasTarget {
  alias_name: string;
  alias_target_id: string; // duplicate of id
  provider_id: string;
  provider_name: string;
  provider_usage_count: number;
  provider_endpoint_id?: string;
  endpoint_url?: string;
  // api_type is no longer returned in detail list
}

export interface ListAliasesParams {
  page?: number;
  page_size?: number;
}

export interface CreateAliasRequest {
  name: string;
}

export interface UpdateAliasRequest {
  name?: string;
  enabled?: boolean;
}

export interface CreateAliasTargetRequest {
  provider_id: string;
  model_id: string;
}

export interface UpdateAliasTargetRequest {
  provider_id?: string;
  model_id?: string;
  enabled?: boolean;
}

// Alias API
export async function listAliases(params: ListAliasesParams = {}): Promise<Alias[]> {
  const query = new URLSearchParams();
  if (params.page !== undefined) query.set("page", params.page.toString());
  if (params.page_size !== undefined) query.set("page_size", params.page_size.toString());

  const queryString = query.toString();
  return requestJson<Alias[]>(`/aliases${queryString ? `?${queryString}` : ""}`);
}

export async function createAlias(payload: CreateAliasRequest): Promise<Alias> {
  return requestJson<Alias>("/aliases", {
    method: "POST",
    body: payload,
  });
}

export async function getAlias(id: string): Promise<Alias> {
  return requestJson<Alias>(`/aliases/${id}`);
}

export async function updateAlias(id: string, payload: UpdateAliasRequest): Promise<Alias> {
  return requestJson<Alias>(`/aliases/${id}`, {
    method: "PUT",
    body: payload,
  });
}

export async function deleteAlias(id: string): Promise<void> {
  await requestJson<void>(`/aliases/${id}`, {
    method: "DELETE",
  });
}

// Alias Target API
export async function listAliasTargetDetails(aliasId: string): Promise<AliasTargetDetail[]> {
  return requestJson<AliasTargetDetail[]>(`/aliases/${aliasId}/targets/details`);
}

export async function createAliasTarget(
  aliasId: string,
  payload: CreateAliasTargetRequest,
): Promise<AliasTarget> {
  return requestJson<AliasTarget>(`/aliases/${aliasId}/targets`, {
    method: "POST",
    body: payload,
  });
}

export async function updateAliasTarget(
  aliasId: string,
  targetId: string,
  payload: UpdateAliasTargetRequest,
): Promise<AliasTarget> {
  return requestJson<AliasTarget>(`/aliases/${aliasId}/targets/${targetId}`, {
    method: "PUT",
    body: payload,
  });
}

export async function deleteAliasTarget(aliasId: string, targetId: string): Promise<void> {
  await requestJson<void>(`/aliases/${aliasId}/targets/${targetId}`, {
    method: "DELETE",
  });
}
